### PR TITLE
[Chat][Feature]Storage Implementation with Tree based data structure

### PR DESCRIPTION
**Pull request by Mohtasim**
## Purpose

* [Storage Implementation with Tree based data structure](https://github.com/Azure/communication-ui-library-android/commit/92cadab10435e0218674e02cf6943482873c2f81)

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Pull Request Type
What kind of change does this Pull Request introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Documentation content changes
- [ ] Other... Please describe:

## Pull Request Checklist

- [ ] Public API changes
- [ ] Verified for cross-platform
- [ ] Synced with iOS, Web team
- [ ] Internal review completed
- [ ] UI Changes
  - [ ] Screen captures included
  - [ ] Tested for Light/Dark mode
  - [ ] Tested for screen rotation
  - [ ] Tested on Tablet and small screen device (5")
  - [ ] Include localization changes
- [ ] Tests
  - [ ] Memory leak analysis performed
  - [ ] Tested on API 21, 26 and latest 
  - [ ] Tested for background mode
  - [ ] Unit Tests Included
  - [ ] UI Automated Tests Included

## How to Test

* Open...
* Click on...

## What to Check
Verify that the following are valid
* ...

### CODE REVIEW

File: azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/repository/storage/MessageStorage.kt [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+        var newMessage = MessageInfoModel(
+            id = previousMessage.id,
+            internalId = previousMessage.internalId,
+            content = message.content,
+            messageType = previousMessage.messageType,
+            version = previousMessage.version,
+            senderDisplayName = previousMessage.senderDisplayName,
+            createdOn = previousMessage.createdOn,
+            editedOn = previousMessage.editedOn,
+            deletedOn = previousMessage.deletedOn,
```
Comment (ShaunaSong) [inline azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/repository/storage/MessageStorage.kt:88]: Please map `isCurrentUser` as well

File: azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/repository/storage/MessageStorage.kt [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+import com.azure.android.communication.ui.chat.models.MessageInfoModel
+import com.azure.android.communication.ui.chat.repository.MessageRepositoryReader
+import com.azure.android.communication.ui.chat.repository.MessageRepositoryWriter
+import com.azure.android.communication.ui.chat.service.sdk.wrapper.ChatMessageType
+import java.util.TreeMap
+import org.threeten.bp.OffsetDateTime
+import org.threeten.bp.ZoneOffset
+
+internal class MessageStorage : MessageRepositoryReader(), MessageRepositoryWriter {
```
Comment (ahammer) [inline azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/repository/storage/MessageStorage.kt:18]: follow the delegate pattern.

MessageRepositoryTreeReader
MessageRepositoryTreeWriter

add a factory constructor to MessageRepository.

Comment (Mohtasim) [reply azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/repository/storage/MessageStorage.kt:18]: Yes. This was an initial draft. Updating before pushing for review.

File: azure-communication-ui/chat/src/test/java/com/azure/android/communication/ui/chat/repository/storage/MessageRepositoryTreeStorageUnitTest.kt [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+import com.azure.android.communication.ui.chat.models.MessageInfoModel
+import com.azure.android.communication.ui.chat.repository.MessageRepository
+import com.azure.android.communication.ui.chat.service.sdk.wrapper.ChatMessageType
+import org.junit.Assert
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.mockito.junit.MockitoJUnitRunner
+
+@RunWith(MockitoJUnitRunner::class)
+class MessageRepositoryTreeStorageUnitTest {
```
Comment (ahammer) [inline azure-communication-ui/chat/src/test/java/com/azure/android/communication/ui/chat/repository/storage/MessageRepositoryTreeStorageUnitTest.kt:15]: We can consolidate MessageRepository tests to cover multiple Storage implementations with one set of tests. Doesn't have to be this PR

File: azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/repository/storage/MessageStorage.kt (hunk 2 above)
Comment (ahammer) [reply azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/repository/storage/MessageStorage.kt:15]: Do we still need this class? It's in the PR, I do see Reader/Writer now.

File: azure-communication-ui/chat/src/test/java/com/azure/android/communication/ui/chat/repository/storage/MessageRepositoryTreeStorageUnitTest.kt (hunk 3 above)
Comment (Mohtasim) [reply azure-communication-ui/chat/src/test/java/com/azure/android/communication/ui/chat/repository/storage/MessageRepositoryTreeStorageUnitTest.kt:15]: I needed to validate if tree based implementations were giving correct output for APIs. 

In my opinion, since the underlaying implementations are separate for each type of implementation. Its safer to have them tested separately rather relying on messageRepo tests. Otherwise messageRepo tests will keep growing and will be hard to follow the coverage of underlaying implementation.

**Review by ahammer — DISMISSED** [review]

File: azure-communication-ui/chat/src/test/java/com/azure/android/communication/ui/chat/repository/storage/MessageRepositoryTreeStorageUnitTest.kt (hunk 3 above)
Comment (ahammer) [reply azure-communication-ui/chat/src/test/java/com/azure/android/communication/ui/chat/repository/storage/MessageRepositoryTreeStorageUnitTest.kt:15]: If the Message Repo API/Contract is doing what it should, the implementation is hidden, that is what we want.

The methods that need testing are still just the things from the reader/writer interfaces, which are exposed via delegates on the message repo.

So a very thorough test suite of MessageRepo is what we want, with all tests passing when running the tests with the chosen storage implementation. That way if we make a new Storage Implementation, you don't need a new test suite, you just tell it "use this implementation" instead and can do TDD to finish it.

Comment (ahammer) [reply azure-communication-ui/chat/src/test/java/com/azure/android/communication/ui/chat/repository/storage/MessageRepositoryTreeStorageUnitTest.kt:15]: I'll resolve for now though, don't want it to hold up this pr

**Review by ShaunaSong — DISMISSED** [review]

**Review by iaulakh — APPROVED** [review]

**Review by ahammer — APPROVED** [review]
